scio-core/src/main/scala/com/spotify/scio/coders/BeamCoders.scala

```diff
@@ -27,100 +28,109 @@ import scala.annotation.tailrec
 /** Utility for extracting [[Coder]]s from Scio types. */
 private[scio] object BeamCoders {
   @tailrec
-  private def unwrap[T](coder: beam.Coder[T]): beam.Coder[T] =
+  private def unwrap[T](options: CoderOptions, coder: beam.Coder[T]): beam.Coder[T] =
     coder match {
-      case c: WrappedCoder[T]       => unwrap(c.bcoder)
-      case c: beam.NullableCoder[T] => c.getValueCoder
-      case _                        => coder
+      case c: MaterializedCoder[T]                            => unwrap(options, c.bcoder)
+      case c: beam.NullableCoder[T] if options.nullableCoders => c.getValueCoder
```

Comment:
what will happen in the case that a user doesn't have the `--nullableCoders` option set, but still uses NullableCoder in their pipeline? (i.e. they have a step like `.map(x => ...)(Coder.beam(NullableCoder.of(...))`

Reply:
The null coder will be propagated. This is required by `countByValue` for example, otherwise we would fail on `null` values.